main.c

```diff
@@ -16,28 +16,53 @@
 
 #include "tinywav.h"
 
-#define NUM_CHANNELS 1
+#include <assert.h>
+
+#define NUM_CHANNELS 2
 #define SAMPLE_RATE 48000
 #define BLOCK_SIZE 512
-#define NUM_ITERATIONS 3
 
 int main(int argc, char *argv[]) {
+  char* outputPath = "output.wav";
+  int totalNumSamples;
+  int framesProcessed = 0;
+  
   if (argc < 2) return -1;
-  const char *outputPath = argv[1];
+  const char *inputPath = argv[1];
 
-  TinyWav tw;
-  tinywav_open_write(&tw, NUM_CHANNELS, SAMPLE_RATE, TW_FLOAT32, TW_INLINE, outputPath);
+  TinyWav twReader, twWriter;
 
-  for (int i = 0; i < NUM_ITERATIONS; i++) {
-    float buffer[BLOCK_SIZE];
-    for (int j = 0; j < BLOCK_SIZE; j++) {
-      buffer[j] = j / ((float) BLOCK_SIZE);
-    }
+  tinywav_open_read(&twReader, inputPath, TW_INLINE);
+  if (twReader.numChannels != NUM_CHANNELS || twReader.h.SampleRate != SAMPLE_RATE) {
+      printf("Supplied test wav has wrong format - should be [%d]channels, fs=[%d]\n", NUM_CHANNELS, SAMPLE_RATE);
+      return -1;
+  }
+  totalNumSamples = twReader.numFramesInHeader;
+  
+  tinywav_open_write(&twWriter, NUM_CHANNELS, SAMPLE_RATE, TW_FLOAT32, TW_INLINE, outputPath);
 
-    tinywav_write_f(&tw, buffer, BLOCK_SIZE);
+  assert(totalNumSamples >= BLOCK_SIZE && "input file needs to be at least 1 block long!");
+  while (framesProcessed < totalNumSamples) {
+    int samplesRead, samplesWritten;
```

Comment:
Is this a timewarp to 1989? C89? Why declare variables upfront like this?

Reply:
I'm indeed stuck in C89  (embedded C project at work), that's why I instinctively did this.
Happy to hear tinywav is C99; a freedom which I will gladly leverage and adjust the code accordingly :-)